src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+use alloc::heap::{EMPTY, allocate, reallocate, deallocate};
+
+// Implemented on top of `core` to demonstrate the ability to substitute HybridVec for Vec
+use core::borrow::{BorrowFrom, ToOwned};
+use core::default::Default;
+use core::fmt;
+use core::hash::{self, Hash};
+use core::intrinsics::forget;
+use core::iter;
+use core::marker::ContravariantLifetime;
+use core::nonzero::NonZero;
+use core::mem;
+use core::num::{Int, UnsignedInt};
+use core::ops;
+use core::ptr;
+use core::raw::Slice as RawSlice;
+use core::usize;
+
+// Make this configurable?
+const ARRAY_SIZE: usize = 32;
+// The number of bytes a type has to be larger than to automatically be allocated on the heap.
+// This limits the on-stack size of `HybridVec` to a maximum of 
+// `ARRAY_SIZE * HEAP_BYTES_THRESHOLD`.
+const HEAP_BYTES_THRESHOLD: usize = 64;
+
+/// An implementation of `HybridVec` that uses stack-allocated storage below a certain length.
+/// A drop-in replacement for `Vec` from the stdlib, with a couple minor changes to the API and internal
+/// behavior.
+pub enum HybridVec<T> {
+    Stack {
+        array: [T; ARRAY_SIZE], 
```

Comment:
That's why I made the performance parameters constants instead of magic numbers. It's meant to be tweaked and tuned, based on what performs the best in the most common cases. I realize 32 is probably too big but I wasn't sure what that number *should* be.

@reem I'm perfectly aware of this limitation. I was hoping DST could help here but I couldn't figure out the syntax wrt `?Sized`. It does switch to heap if the type is too big but that logic is a bit flawed anyways because Rust is going to allocate the stack space for it regardless. I'll probably drop that bit since it doesn't save anyone anything.